gcp/api/cursor.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""OSV API server cursor implementation."""
+
+import base64
+from enum import Enum
+from typing import Self
+import typing
+
+from google.cloud import ndb
+import google.cloud.ndb.exceptions as ndb_exceptions
+
+_FIRST_PAGE_TOKEN = base64.urlsafe_b64encode(b'FIRST_PAGE_TOKEN').decode()
+# Use ':' as the separator as it doesn't appear in urlsafe_b64encode
+# (which is what is used for both _FIRST_PAGE_TOKEN, and ndb.Cursor.urlsafe())
+_METADATA_SEPARATOR = ':'
+
+
+class _QueryCursorState(Enum):
+  """
+  Stores the 3 states a query cursor can be in.
+  """
+
+  # The cursor has reached the end, no need to return to the user
+  ENDED = 0
+  # The cursor is at the very start of the query, no cursor needs to be
+  # set when querying ndb
+  STARTED = 1
+  # ndb.Cursor is set and in progress
+  IN_PROGRESS = 2
+
+
+class QueryCursor:
+  """
+  Custom cursor class that wraps the ndb cursor.
+  Allows us to represent the "starting" cursor.
+
+  The default state is ENDED with no ndb.Cursor.
+
+  This type could have 3 states encoded in _QueryCursorState.
+  If the current state is IN_PROGRESS, self.cursor will not be None.
+  """
+
+  cursor: ndb.Cursor | None = None
+  cursor_state: _QueryCursorState = _QueryCursorState.ENDED
+  query_number: int = 1
```

Comment:
Why not 0 as the default value?

Reply:
I'm guessing this is because the first query is `query_number = 1`? Count from zero or count from one?

Reply:
It's a bit difficult to document why it's 1 here in a comment. It essentially makes another completely separate part of the code a bit easier to write. 

We call `context.should_skip_query()` every time we do a query, and we also want to increment the context.query_counter by 1 no matter what should_skip_query() returns, so we do the increment before calling should_skip_query()

`should_skip_query()` also checks the query_counter, so rather than having - 1 in a few separate places, I just made the query number start at 1.

Reply:
Now looking at it, maybe I should move the query_counter increment into should_skip_query(), since they are always paired together, but it also feels weird to be mutating the state in a "question" method.

Reply:
IMO we shouldn't be coupling implementation details of another part of the code with design decisions here. 

+1 to moving the query_counter increment to `should_skip_query`. This way we also avoid the -1 right? 

Re naming, we could rename it to something like `preprocess_and_check_subquery()`, and invert the return value so it'll look something like:


```python
if not context.preprocess_and_check_subquery():
  return []
```

WDYT?

Reply:
Yeah I'll do that!

The most "correct" version of this is probably a function decorator which does the skipping and counting, but I think that disconnects the .iter() too much from the increment of the query counter.

Reply:
Hmm it'll avoid the -1 in that specific instances, but we'll still need -1 for two other places. I think it'll be clearer to define the query number as incrementing once before the start of the query, so everything that uses the query number can know that the first query is numbered "1". Rather than having inconsistent query number matching. I'll add that to the comments to make it clearer. 

I'll make the change to combine the +1 with the skip check though. 

In fact making this change made me aware of the need to check the query number when retrieving the ndb cursor which I am not currently doing.

Reply:
After implementing it, I think it's still much neater/clearer to keep it as two separate lines, as it avoids mutation happening inside an if statement, and clearly shows that the query number increments before anything related to the query begins.

Reply:
SGTM